example/src/main/java/io/netty/example/dns/doh/DoHClient.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.dns.doh;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+import io.netty.handler.codec.dns.DefaultDnsQuery;
+import io.netty.handler.codec.dns.DefaultDnsQuestion;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+
+
+import java.net.URL;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+public final class DoHClient {
+
+    private static final String QUERY_DOMAIN = "www.example.com";
+
+    private DoHClient() {
+    }
+
+    public static void main(String[] args) throws Exception {
+        URL url = new URL("https://dns.google/dns-query");
+        EventLoopGroup group = new NioEventLoopGroup();
+
+        try {
+            final SslContext sslContext = SslContextBuilder.forClient()
+                    .protocols("TLSv1.2", "TLSv1.3")
+                    .applicationProtocolConfig(new ApplicationProtocolConfig(
+                            ApplicationProtocolConfig.Protocol.ALPN,
+                            ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                            ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                            ApplicationProtocolNames.HTTP_2,
+                            ApplicationProtocolNames.HTTP_1_1))
+                    .build();
+
+            Bootstrap b = new Bootstrap();
+            b.group(group)
+                    .channel(NioSocketChannel.class)
+                    .handler(new Initializer(sslContext, url));
+
+            final Channel ch = b.connect(url.getHost(), url.getDefaultPort()).sync().channel();
+
+            Thread.sleep(1000); // Wait for TLS Handshake to finish
```

Comment:
don't use a sleep here, use the handshake future.

```
sslHandler.handshakeFuture().addListener(future -> {
                            if (future.isSuccess()) {
```

Reply:
Tried `ch.pipeline().get(SslHandler.class).handshakeFuture().sync();` as mentioned here: https://stackoverflow.com/a/44714941/8267040

But got exception:

```
08 May 2020 03:35:15.183 [nioEventLoopGroup-2-1] DEBUG i.n.u.Recycler - -Dio.netty.recycler.maxCapacityPerThread: 4096
08 May 2020 03:35:15.183 [nioEventLoopGroup-2-1] DEBUG i.n.u.Recycler - -Dio.netty.recycler.maxSharedCapacityFactor: 2
08 May 2020 03:35:15.183 [nioEventLoopGroup-2-1] DEBUG i.n.u.Recycler - -Dio.netty.recycler.linkCapacity: 16
08 May 2020 03:35:15.184 [nioEventLoopGroup-2-1] DEBUG i.n.u.Recycler - -Dio.netty.recycler.ratio: 8
08 May 2020 03:35:15.193 [nioEventLoopGroup-2-1] DEBUG i.n.b.AbstractByteBuf - -Dio.netty.buffer.checkAccessible: true
08 May 2020 03:35:15.193 [nioEventLoopGroup-2-1] DEBUG i.n.b.AbstractByteBuf - -Dio.netty.buffer.checkBounds: true
08 May 2020 03:35:15.194 [nioEventLoopGroup-2-1] DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@3ba116c7
08 May 2020 03:35:15.504 [main] DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@1c80e49b
08 May 2020 03:35:15.514 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x74ce20d4, L:/10.100.100.3:22689 - R:dns.google/8.8.8.8:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
Exception in thread "main" io.netty.handler.codec.UnsupportedMessageTypeException: io.netty.handler.codec.dns.DefaultDnsQuery (expected: io.netty.buffer.ByteBuf)
	at io.netty.handler.ssl.SslHandler.write(SslHandler.java:762)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
08 May 2020 03:35:17.630 [nioEventLoopGroup-2-1] DEBUG i.n.b.PoolThreadCache - Freed 13 thread-local buffer(s) from thread: nioEventLoopGroup-2-1
```

Reply:
Adding `Thread.sleep(500);` after `ch.pipeline().get(SslHandler.class).handshakeFuture().sync();` solves the issue. 
Not sure what's happening. ;/

Reply:
@normanmaurer Can you help here? I tried a lot but can't get it working. `handshakeFuture().sync()` is not working as expected.

Reply:
@hyperxpro that exception pretty much speaks for itself, DefaultDnsQuery is not being "encoded" into ByteBuf before it reaches SslHandler, check your pipeline configuration / order of encoders / decoders.

Reply:
Checked again. The pipeline and order are alright. ALPN Handler is actually taking the time to add correct handlers. Once Handshake is finished, the ALPN handler negotiates protocol and adds HTTP/2 or HTTP/1.1 handler along with DoH encoder and decoder.

With 500 ms. wait:
```
DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (Initializer$ALPNHandler#0 = testing.Initializer$ALPNHandler)}
10 May 2020 16:05:07.521 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0xad8e9662, L:/10.100.100.3:27267 - R:dns.google/8.8.8.8:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:05:07.672 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0xad8e9662, L:/10.100.100.3:27267 - R:dns.google/8.8.8.8:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (HttpToHttp2ConnectionHandler#0 = io.netty.handler.codec.http2.HttpToHttp2ConnectionHandler), (HttpObjectAggregator#0 = io.netty.handler.codec.http.HttpObjectAggregator), (DoHResponseDecoder#0 = testing.DoHResponseDecoder), (DoHQueryEncoder#0 = testing.DoHQueryEncoder), (Handler#0 = testing.Handler)}
```

Without 500ms wait:
```
DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (Initializer$ALPNHandler#0 = testing.Initializer$ALPNHandler)}
DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (Initializer$ALPNHandler#0 = testing.Initializer$ALPNHandler)}
10 May 2020 16:05:47.614 [main] DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@470a696f
10 May 2020 16:05:47.623 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x0897fa2a, L:/10.100.100.3:27308 - R:dns.google/8.8.8.8:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
Exception in thread "main" io.netty.handler.codec.UnsupportedMessageTypeException: io.netty.handler.codec.dns.DefaultDnsQuery (expected: io.netty.buffer.ByteBuf)
	at io.netty.handler.ssl.SslHandler.write(SslHandler.java:762)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
10 May 2020 16:05:49.722 [nioEventLoopGroup-2-1] DEBUG i.n.b.PoolThreadCache - Freed 13 thread-local buffer(s) from thread: nioEventLoopGroup-2-1
```

Reply:
Added log4j2 logger and checked timestamps and `ch.writeAndFlush(query).awaitUninterruptibly(10, TimeUnit.SECONDS);` instead of `ch.writeAndFlush(query).sync();`

With 500ms wait (Works perfectly):
```
10 May 2020 16:41:47.554 [main] DEBUG t.DoHClient - Handshake Finished, out of `sync()`
10 May 2020 16:41:47.555 [main] DEBUG t.DoHClient - Pipelines: DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (Initializer$ALPNHandler#0 = testing.Initializer$ALPNHandler)}
10 May 2020 16:41:47.561 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x30e9cb67, L:/10.100.100.3:28329 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:41:47.644 [nioEventLoopGroup-2-1] DEBUG t.DoHQueryEncoder - Initialized
10 May 2020 16:41:47.651 [nioEventLoopGroup-2-1] DEBUG t.I.ALPNHandler - Using HTTP/2 (h2)
10 May 2020 16:41:47.717 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x30e9cb67, L:/10.100.100.3:28329 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:41:48.057 [main] DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@470a696f
10 May 2020 16:41:48.068 [nioEventLoopGroup-2-1] DEBUG t.DoHQueryEncoder - Received DnsQuery for encoding: DefaultFullHttpRequest(decodeResult: success, version: HTTP/1.1, content: PooledUnsafeDirectByteBuf(ridx: 0, widx: 33, cap: 256))
POST /dns-query HTTP/1.1
host: dns.google
content-type: application/dns-message
accept: application/dns-message
x-http2-scheme: https
10 May 2020 16:41:48.084 [main] DEBUG t.DoHClient - Write successful for DefaultDnsQuery, Pipelines: DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (HttpToHttp2ConnectionHandler#0 = io.netty.handler.codec.http2.HttpToHttp2ConnectionHandler), (HttpObjectAggregator#0 = io.netty.handler.codec.http.HttpObjectAggregator), (DoHResponseDecoder#0 = testing.DoHResponseDecoder), (DoHQueryEncoder#0 = testing.DoHQueryEncoder), (Handler#0 = testing.Handler)}
name: www.example.com.
93.184.216.34
```

Without 500ms wait (DoHQueryEncoder is initialized but not receiving any DnsQuery):
```
10 May 2020 16:43:28.854 [main] DEBUG t.DoHClient - Handshake Finished, out of `sync()`
10 May 2020 16:43:28.855 [main] DEBUG t.DoHClient - Pipelines: DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (Initializer$ALPNHandler#0 = testing.Initializer$ALPNHandler)}
10 May 2020 16:43:28.858 [main] DEBUG i.n.u.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@73c60324
10 May 2020 16:43:28.862 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x149dbe03, L:/10.100.100.3:28388 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:43:28.938 [nioEventLoopGroup-2-1] DEBUG t.DoHQueryEncoder - Initialized
10 May 2020 16:43:28.944 [nioEventLoopGroup-2-1] DEBUG t.I.ALPNHandler - Using HTTP/2 (h2)
10 May 2020 16:43:28.947 [main] DEBUG t.DoHClient - Write successful for DnsQuery, Pipelines: DefaultChannelPipeline{(SslHandler#0 = io.netty.handler.ssl.SslHandler), (HttpToHttp2ConnectionHandler#0 = io.netty.handler.codec.http2.HttpToHttp2ConnectionHandler), (HttpObjectAggregator#0 = io.netty.handler.codec.http.HttpObjectAggregator), (DoHResponseDecoder#0 = testing.DoHResponseDecoder), (DoHQueryEncoder#0 = testing.DoHQueryEncoder), (Handler#0 = testing.Handler)}
10 May 2020 16:43:29.010 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x149dbe03, L:/10.100.100.3:28388 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:43:40.964 [nioEventLoopGroup-2-1] DEBUG i.n.b.PoolThreadCache - Freed 15 thread-local buffer(s) from thread: nioEventLoopGroup-2-1
```

Reply:
Just wondering, why I get 2 Handshake message?
```
10 May 2020 16:43:28.862 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x149dbe03, L:/10.100.100.3:28388 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
10 May 2020 16:43:29.010 [nioEventLoopGroup-2-1] DEBUG i.n.h.s.SslHandler - [id: 0x149dbe03, L:/10.100.100.3:28388 - R:dns.google/8.8.4.4:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
```

Reply:
@normanmaurer Eagerly waiting for your response.

Reply:
I'll take a look this afternoon.

Reply:
io.netty.handler.codec.dns.DoHQueryEncoder fails validation plugin btw.

Reply:
```
Index: codec-dns/src/main/java/io/netty/handler/codec/dns/DoHQueryEncoder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- codec-dns/src/main/java/io/netty/handler/codec/dns/DoHQueryEncoder.java	(revision 3f5a82f729f41b5e626e7cb3039c0b503dbcb05e)
+++ codec-dns/src/main/java/io/netty/handler/codec/dns/DoHQueryEncoder.java	(date 1589116490359)
@@ -36,8 +36,8 @@
 public class DoHQueryEncoder extends MessageToMessageEncoder<DnsQuery> {
 
     private final DnsQueryEncoder encoder;
-    private final boolean HTTP2;
-    private final boolean GET;
+    private final boolean http2;
+    private final boolean get;
     private final URL url;
 
     /**
@@ -51,14 +51,14 @@
     }
 
     /**
-     * Creates a new encoder with {@linkplain DnsRecordEncoder#DEFAULT the default record encoder}, uses HTTP POST method
-     * and specifies if we're using HTTP/2 (h2).
+     * Creates a new encoder with {@linkplain DnsRecordEncoder#DEFAULT the default record encoder}, uses HTTP POST
+     * method and specifies if we're using HTTP/2 (h2).
      *
-     * @param HTTP2 Use HTTP/2 (h2)
+     * @param http2 Use HTTP/2 (h2)
      * @param url   DoH Upstream Server
      */
-    public DoHQueryEncoder(boolean HTTP2, URL url) {
-        this(DnsRecordEncoder.DEFAULT, HTTP2, false, url);
+    public DoHQueryEncoder(boolean http2, URL url) {
+        this(DnsRecordEncoder.DEFAULT, http2, false, url);
     }
 
     /**
@@ -66,14 +66,14 @@
      * {@code GET} and {@code url}
      *
      * @param recordEncoder DNS Record Encoder
-     * @param HTTP2         Use HTTP/2 (h2)
-     * @param GET           Use HTTP GET method
+     * @param http2         Use HTTP/2 (h2)
+     * @param get           Use HTTP GET method
      * @param url           DoH Upstream Server
      */
-    public DoHQueryEncoder(DnsRecordEncoder recordEncoder, boolean HTTP2, boolean GET, URL url) {
+    public DoHQueryEncoder(DnsRecordEncoder recordEncoder, boolean http2, boolean get, URL url) {
         this.encoder = new DnsQueryEncoder(recordEncoder);
-        this.HTTP2 = HTTP2;
-        this.GET = GET;
+        this.http2 = http2;
+        this.get = get;
         this.url = url;
     }
 
@@ -83,7 +83,7 @@
         encoder.encode(msg, byteBuf);
 
         FullHttpRequest fullHttpRequest;
-        if (GET) {
+        if (get) {
             fullHttpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET,
                     url.getPath() + "?dns=" + Base64.encode(byteBuf).toString(Charset.forName("UTF-8")));
         } else {
@@ -98,7 +98,7 @@
                 .add(HttpHeaderNames.ACCEPT, "application/dns-message");
 
         // If we're using HTTP/2 (h2) then we'll add "x-http2-scheme" header
-        if (HTTP2) {
+        if (http2) {
             fullHttpRequest.headers().add(HttpConversionUtil.ExtensionHeaderNames.SCHEME.text(),
                     HttpScheme.HTTPS.name());
         }
```

Reply:
Will fix it at final stage.

Reply:
@hyperxpro as I mentioned earlier, it's because of your pipeline, there is a race which is why the sleep (sometimes) works.. sync on handshakeFuture completes and the main thread continues before the event loop configures the pipeline here io.netty.example.dns.doh.Initializer.ALPNHandler#configurePipeline

Reply:
How to wait for ALPNHandler to finish?

Reply:
here is the entry point of where the alpn handler configures the pipeline io.netty.handler.ssl.ApplicationProtocolNegotiationHandler#userEventTriggered which is more or less at the same point of where you stop blocking the main thread.

Reply:
I know this is lame but never used `userEventTriggered`. Can you help me with an example or some reference?

Reply:
create a promise, pass it into io.netty.example.dns.doh.Initializer then block on that instead of handshakeFuture, the promise you pass into Initializer should be completed after you modify the pipline in io.netty.example.dns.doh.Initializer.ALPNHandler#configurePipeline

Reply:
also for what it is worth "TLSv1.3" caused my JVM to fail.. remember Netty CI builds against 8 etc too.

java.lang.IllegalArgumentException: TLSv1.3
	at sun.security.ssl.ProtocolVersion.valueOf(ProtocolVersion.java:187)
	at sun.security.ssl.ProtocolList.convert(ProtocolList.java:84)
	at sun.security.ssl.ProtocolList.<init>(ProtocolList.java:52)
	at sun.security.ssl.SSLEngineImpl.setEnabledProtocols(SSLEngineImpl.java:2070)
	at io.netty.handler

Reply:
`ChannelPromise channelPromise = ch.newPromise();`

How can I pass this promise into Initializer before channel is connected?

Reply:
https://gist.github.com/johnou/61129b6fae2f3df8517a417f86097db4

Reply:
@hyperxpro try taking at the second version, I updated it just now.

Reply:
https://github.com/hyperxpro/netty/blob/25872ac26d998a6183d5609dcf2a7c4c29db3d11/codec-dns/src/main/java/io/netty/handler/codec/dns/DoHQueryEncoder.java#L82 don't need to be freed because releasing it will cause `io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1`.